src/Database/LSMTree/Internal/Vector/Growing.hs

```diff
@@ -0,0 +1,117 @@
+                             !(PrimVar s Int)         -- Reference to the length
+
+-- | Creates a new, initially empty, vector.
+new :: Int                      -- ^ Initial buffer size
+    -> ST s (GrowingVector s a) -- ^ Construction of the vector
+new illegalInitialBufferSize | illegalInitialBufferSize <= 0
+    = error "Initial buffer size not positive"
+new initialBufferSize
+    = do
+        buffer <- Mutable.new initialBufferSize
+        bufferRef <- newSTRef buffer
+        lengthRef <- newPrimVar 0
+        return (GrowingVector bufferRef lengthRef)
+
+{-|
+    Appends a value a certain number of times to a vector. If a negative number
+    is provided as the count, the vector is not changed.
+-}
+append :: forall s a . GrowingVector s a -> Int -> a -> ST s ()
+append _ pseudoCount _ | pseudoCount <= 0
+    = return ()
+append (GrowingVector bufferRef lengthRef) count val
+    = do
+          length <- readPrimVar lengthRef
+          makeRoom
+          buffer' <- readSTRef bufferRef
+          Mutable.set (Mutable.slice length count buffer') val
+    where
+
+    makeRoom :: ST s ()
+    makeRoom = do
+        length <- readPrimVar lengthRef
+        when (count > maxBound - length) (error "New length too large")
+        buffer <- readSTRef bufferRef
+        let
+
+            bufferSize :: Int
+            !bufferSize = Mutable.length buffer
+
+            length' :: Int
+            !length' = length + count
+
+        when (bufferSize < length') $ do
+            let
+
+                higherBufferSizes :: [Int]
+                higherBufferSizes = tail (init ++ [last]) where
+
+                    init :: [Int]
+                    last :: Int
+                    (init, last : _) = span (<= maxBound `div` 2) $
+                                       iterate (* 2) bufferSize
+                    {-NOTE:
+                        In order to prevent overflow, we have to start with the
+                        current buffer size here, although we know that it is
+                        not sufficient.
+                    -}
+
+                sufficientBufferSizes :: [Int]
+                sufficientBufferSizes = dropWhile (< length') higherBufferSizes
+
+            case sufficientBufferSizes of
+                []
+                    -> error "No sufficient buffer size available"
+                bufferSize' : _
+                    -> Mutable.grow buffer (bufferSize' - bufferSize) >>=
+                       writeSTRef bufferRef
+        writePrimVar lengthRef length'
+
+{-|
+    Turns a growing vector into an ordinary vector, thereby invalidating the
+    growing vector. Executing @unsafeFreeze vec@ is only safe when @vec@ is not
+    used afterwards.
+-}
+unsafeFreeze :: GrowingVector s a -> ST s (Vector a)
+unsafeFreeze (GrowingVector bufferRef lengthRef) = do
+    buffer <- readSTRef bufferRef
+    length <- readPrimVar lengthRef
+    force <$> Mutable.unsafeFreeze (Mutable.take length buffer)
```

Comment:
If I remember correctly, `force` will always copy the contents into a new vector. As such, this implementation is actually a safe `freeze`

Reply:
I had wondered about this already before sending you this pull request. I have checked now: From a quick look at the code, I get indeed the impression that `force` will always copy the contents. However, the API documentation says this:

> Yield the argument, but force it not to retain any extra memory, possibly by copying it.

So the specification leaves open the possibility of `force` returning the original vector if this vector uses its underlying array in full. We might hope that future versions of the `vector` package won’t make use of this opportunity, but I’d like to err on the safe side and continue considering the above function unsafe.

Reply:
I've created https://github.com/haskell/vector/issues/511 about the specification of `force`.

Different suggestion: use `freeze` instead of `force`+`unsafeFreeze`. The result is the same, and it's less ambiguous whether there is a copy or not. And then we can call the function `freeze` instead of `unsafeFreeze`

Reply:
Indeed, `freeze` should always generate a copy, and, while its documentation doesn’t specifically mention that it copies only the relevant slice, it would be weird for it not to do so, and the current implementation seems to do it. I’ll change the code to use `freeze` instead of `force` and `unsafeFreeze`.

Reply:
Done in 49eb151ca766b6c6a33685753b715bd4b1d304d2.